scl/paloalto/panos.conf

```diff
@@ -0,0 +1,100 @@
+#############################################################################
+# Copyright (c) 2020 Balabit
+# Copyright (c) MileK <mileek@gmail.com>
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 2 as published
+# by the Free Software Foundation, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# As an additional exemption you are allowed to compile & link against the
+# OpenSSL libraries as published by the OpenSSL project. See the file
+# COPYING for details.
+#
+#############################################################################
+#
+#scl/paloalto/panos.conf -- Paloalto PAN-OS parser fro syslog-ng
+
+block parser panos-parser (prefix(".panos.") source("syslog")...) {
```

Comment:
Just sharing an idea, act upon on this if you feel like.
Your parser now one big parser, that does what it needs to, but imho it would be nice to split it up into different section for easier reading. You have two basic section, one handling common part of the log parser, and the different emitter specific one. So you could create a `panos-common-parser`, and a `panos-application-specific` parser. The later would deal with the big block of IFs.

```
block parser panos-parser(...)
{
   panos-common-parser(...);
   panos-app-specific-parser();
}

block parser panos-app-specific-parser(...)
{
...
# Common fields 
    parser {
      csv-parser(
        columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time")
      delimiters(',')
      prefix("`prefix`")
     );
   };
...
};


block parser panos-common-parser(...)
{
# Parse logs according to panos.type field
    if (match('SYSTEM', value('.panos.type'))) {
    ...
};
```

The other idea that could help with reading and maintaining this parser would be to use application parser/appmodel.
Instead of listing all of those IFs, you could just say that you want all of the parsers that has certain topic:
```
block parser panos-parser(...)
{
   panos-common-parser(...);
   app-parser(topic(panos-application));
}
```
In this example a new topic: `panos-application`. This parser going to try all of the `panos-application` parsers.
The question is how syslog-ng could know which is a `panos-application` parser, and the answer you must specify them:

```
application panos-system[panos-application] {
  filter { match('SYSTEM', value('.panos.type'); };
  parser {
     csv-parser(
        columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time")
      delimiters(',')
      prefix("`prefix`")
     );
 };
};

application panos-config[panos-application] {
  filter { match('CONFIG', value('.panos.type'); };
  parser {
csv-parser(
columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time","host_name","vsys","command","admin","client","result","configuration_path","sequence_number","action_flags","before_change_detail","after_change_detail","devicegroup_level1","devicegroup_level2","devicegroup_level3","devicegroup_level4","vsys_name","dvc_name")
        prefix("`prefix`")
        delimiters(',')
        ); 
 };
};
...
```

With the `application some-name[topic-to-use-this-parser]` you can extend the list of parser easily.


Oh and btw, there are already topics for the `default-network-drivers` which you could check out for examples, and actually also include to that your parser probably `syslog-raw` would be a good candidate:

```
application panos[syslog-raw] {
    filter { # here you should put something that is uniq for panos logs, i do not have the knowhow for that
    };
   parser { panos-parser(); }
};

Reply:
Hi Peter,
I forgot to introduce `prefix` variable into match `value` so after introducing my value match does not work anymore.
`filter { match("SYSTEM", value("`prefix`type"));};`
 I guess since `application` does not belong to panos-parser block.
It could be that I misunderstood your post but then again I could not find documentation in regards to either `app-parser` nor `application`.

E.g.
```
block parser panos-parser (prefix(".panos.") ...) {
  channel {
    rewrite {
      set("${LEGACY_MSGHDR}${MESSAGE}" value("MESSAGE"));
      set("paloalto_panos", value("PROGRAM"));
     };
# Common fields - set dot-nv-pairs
    parser {
      csv-parser(
        columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time")
        prefix("`prefix`")
        delimiters(',')
      );
    };
  };
  app-parser(topic(panos-application));
};

# Parse logs according to "type" field
# SYSTEM
application panos-system[panos-application] {
  filter { match("SYSTEM", value("`prefix`type"));};
  parser {
     csv-parser(
        columns("future_use1","receive_time","serial_number","type","log_subtype","version","generated_time","vsys","event_id","object","future_use3","future_use4","module","severity","description","sequence_number","action_flags","devicegroup_level1","devicegroup_level2","devicegroup_level3","devicegroup_level4","vsys_name","dvc_name")
      delimiters(',')
      prefix("`prefix`")
     );
 };
};
```

Reply:
We discussed about this with the team, and concluded that we do not want to increase the number of app-parser topics at this point. On the other hand, `panos-parser` would nicely fit in the `syslog` topic.

> [...] I could not find documentation in regards to either `app-parser` nor `application`.

Yes, this is a bit unfortunate, that `app-parser` is not yet documented. It is relatively recent addition to syslog-ng, a bit experimental. On the other hand, it is very convenient feature from user experience point of view, so we would like to ask you to add `panos-parser()` to the syslog topic.

All you need to do is to add

```
application panos[syslog] {
    filter { message(".*,paloalto.*"); };
    parser { panos-parser(); };
};
```

to the end of the the scl file. 

You can test it with:

```
@version: 3.26

@include "scl.conf"

log {
  source { network(port(5555)); };
  parser { app-parser(topic(syslog)); };
  destination { file("/dev/stdout" template("$(format-json --scope all-nv-pairs)\n")); };
};
```

and by sending an example message:

```
echo "<14>Apr 14 16:54:18 paloalto.test.net 1,2020/04/14 16:54:18,unknown,CONFIG,0,0,2020/04/14 16:54:18,10.0.10.55,,set,admin,Web,Succeeded, deviceconfig system,127,0x0,0,0,0,0,,paloalto" | nc localhost 5555
```

The result is the same when you explicitely added `panos-parser` to the config, with a little extra: a new name value also appeared in the message besides the usual:

```
  "_app": {
    "name": "panos"
  },
```

Reply:
> We discussed about this with the team, and concluded that we do not want to increase the number of app-parser topics at this point.
Why?

Reply:
I didn't see too much value in allowing users to extend the set of PanOS log types supported. If something is not, it should be added to the panos-parser() and not into local configuration files.